### ISSUE TITLE
Random LTP test fails with "IPC process-exit notification fails" or "Failed to send IPC msg" or "Failed to exchange the secret key via RPC"

### ISSUE THREAD

**Issue by anjalirai-intel** [oscarlab/graphene#2531]
## Description of the problem
For random tests we are seeing the following errors after the tests have finished and the return code for the test is 0.
Graphene-Native:
**waitpid03**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P2355:T25:waitpid03] error: Failed to send IPC msg to 2307: -32
[P2355:T25:waitpid03] error: Sending IPC process-exit notification failed: -13
```

**waitpid04**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P2361:T2:waitpid04] error: Sending IPC process-exit notification failed: -13
```
**fcntl12_64**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P80670:T2:fcntl12_64] error: Failed to send IPC msg to 80668: -32
```
**recv01**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P34450:T2:recv01] error: Sending IPC process-exit notification failed: -13
```

**recvfrom01**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P34458:T2:recvfrom01] error: Sending IPC process-exit notification failed: -13
```

Graphene-SGX:
**sendfile04**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: Failed to exchange the secret key via RPC: -21
error: Failed to initialize child process: -21
```

**fork08**
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: Failed to exchange the secret key via RPC: -21
error: Failed to initialize child process: -21
```

```
<testcase classname="apps.LTP" name="fork08" time="4.398">
    <system-out>fork08      0  TINFO  :  parent forksval: 1
fork08      0  TINFO  :  second child got char: b
fork08      1  TPASS  :  Test passed in childnumber 2
fork08      0  TINFO  :  parent forksval: 2
fork08      0  TINFO  :  exit status of wait  expected 0 got 0
fork08      1  TPASS  :  parent test PASSED
fork08      0  TINFO  :  exit status of wait  expected 0 got 0
fork08      2  TPASS  :  parent test PASSED
fork08      0  TINFO  :  exit status of wait  expected 0 got 0
fork08      3  TPASS  :  parent test PASSED
fork08      0  TINFO  :  Number of processes forked is 2
</system-out>
    <system-err>error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: Failed to exchange the secret key via RPC: -21
error: Failed to initialize child process: -21
</system-err>
    <properties>
      <property name="returncode" value="0"/>
    </properties>
    <failure message="Error Message=error: Failed to exchange the secret key via RPC: -21"/>
  </testcase>
```

**Stress-Ng Test:**

1. graphene-direct stress-ng --kill 0 --timeout 60s--verbose
2. graphene-direct stress-ng --nice 0 --timeout 60s --verbose

## Steps to reproduce
1. Run LTP tests
2. For Stress-Ng tests failures
apt-get install stress-ng (Install stress -ng first and then run the tests)

## Tests
sendfile04, waitpid03, waitpid04, fcntl12_64, recv01, recvfrom01, fork08, send01

**Comment by dimakuv** [comment]
This is dependent on https://github.com/gramineproject/gramine/issues/21. Looks like the same issue.

**Comment by boryspoplawski** [comment]
@dimakuv are you sure? I thought LTP tests spawn child processes to do actual testing and the parent (which is an IPC leader) waits for their finish.

**Comment by dimakuv** [comment]
@boryspoplawski No, LTP does a lot of weird (and not entirely correct) stuff in many tests.

Here is just one example (I looked into the first reported test here, `waitpid03`, and bingo):
```
// excerpt from waitpid03.c
/*
 * ALGORITHM
 *  Parent forks numerous (22 = MAXUPRC - 3) children, and starts waits :
 *  Should only wait for the specific child, a second wait on the same
 *  child should return with -1 and not one of the other zombied
 *  children.
 */

int main() {
        while (++ikids < MAXUPRC) {
            pid[ikids] = FORK_OR_VFORK();
        }

        /* Wait on one specific child */
        ret = waitpid(pid[MAXUPRC / 2], &status, 0);

    cleanup();
}

static void cleanup(void)
{
    while (ikids-- > 1)
        kill(pid[ikids], SIGKILL);
}
```

As you can see, the parent creates a bunch of child processes, then waits for **one** child process, then sends `kill(SIGKILL)` to all other kids, and **doesn't bother waiting**.

**Comment by dimakuv** [comment]
> I thought LTP tests spawn child processes to do actual testing and the parent (which is an IPC leader) waits for their finish.

Actually, I re-read your comment more carefully. And what you're saying here is that there is a "init-like" LTP process that spawns the child process, and the actual `main()` stuff happens in that child process. No, I don't remember anything like this in LTP. Are you confusing with that-other-test-suite?

**Comment by boryspoplawski** [comment]
> Actually, I re-read your comment more carefully. And what you're saying here is that there is a "init-like" LTP process that spawns the child process, and the actual `main()` stuff happens in that child process. No, I don't remember anything like this in LTP. Are you confusing with that-other-test-suite?

Yes, that's what I meant. 
It's probably different for different LTP tests, I think they changes some stuff around. After checking: yes, some tests run directly, some spawn a child process which does the testing (the newer interface does the latter).

**Comment by vijaydhanraj** [comment]
Was looking at issue #2426(waitpid03 ltp failure test) but since the issue is discussed here, will add my comments.

Found `waitpid03` test to be doing something like below,
1. Parent forks around 24 (pids ranging from 1-25) threads and all the children are paused.
2. Then the parent sends `SIGTERM` signal to all the children to exit gracefully.
3. Parent waits for one particular child to exit.
4. Only on failure or as a part of the cleanup code does the parent sends `SIGKILL` signal.

Initial Analysis:

This issue is more related to the IPC communication between the leader and child rather than `waitpid` syscall itself.

IIUC, the general IPC flow between the leader and the child when a `SIGTERM` signal is received is as follows,
1.	Leader sends IPC_MSG_PID_KILL(1, 1, 2, *15*) to the child and waits for a response.
2.	Child receives the message and calls the appropriate callback (in this case, IPC_MSG_PID_KILL) and sends an ack response back.
3.	Now after the job is done, the child sends an IPC message to the leader and waits for an ack response.
4.	Leader clears `POSIX locks` and sends an ack response message to the child.
5.	Child sends IPC_MSG_CHILDEXIT(1, 2, 0, 15) to the leader, closes handles, and waits for an ack response from the leader.
6.	Leader now knows the child died and calls the appropriate callback.
7.	Child sends another request to the leader to `release_id_range`
8.	Leader calls the appropriate callback and child exits with `exit code 143` which tells that SIGTERM was caught and handled.
From the logs that I collected (PS: the repro is a little difficult and need to run more iterations to hit issue), I see that the child(`T24, PID 39264`) that is awaiting a response at step#3 receives the following error message,
```
[P39264:shim] error: IPC worker: receiving message header from 39170 failed: -13
[P39264:shim] error: IPC worker: failed to receive an IPC message from 39170: -13
```
Although the leader received the message, the child doesn’t know it yet and dumps the following error message,
```
[P39264:T24:waitpid03] debug: Waiting for a response to 2
[P39264:shim] debug: IPC leader disconnected
.
.
[P39264:shim] debug: Unknown process (vmid: 0x9902) disconnected ==> 0x9902 is the leader(39170)
.
.
39264:shim] debug: Woke up a thread waiting for a message from a disconnected process
[P39264:T24:waitpid03] warning: IPC recipient *39170* died while we were waiting for a message response
```
While all this is happening, the leader is continuing to service other children and it had sent another IPC_MSG_PID_KILL(1, 1, *25*, 15) to `T25` and ends up in the same pattern.
One interesting observation is that the parent’s signal queue is blocked, and the following log is thrown up,
```
[P39170:shim] debug: Signal 17 queue of the process is full, dropping incoming signal
``` 

Attaching my log for reference. Lines `3747-8844` show us interesting info regarding the crash.

Few things to debug:
1.	Why does the child fail to receive IPC from the parent suddenly although the parent is still active?
2.	Need to debug this error message “Signal 17 queue of the process is full, dropping incoming signal” as this could be the reason the parent failed in the first place. 


[run1_waitpid03.txt](https://github.com/oscarlab/graphene/files/6863832/run1_waitpid03.txt)

**Comment by vijaydhanraj** [comment]
Given the above analysis, @boryspoplawski @dimakuv this seems to be a genuine issue. What do you think?

**Comment by boryspoplawski** [comment]
@vijaydhanraj are you sure your analysis is correct? To be specific: the part about IPC message receiving failure. Are you sure it happens while IPC leader is still alive?
From a quick look at the code it seems that we've already correctly identified the culprit as https://github.com/gramineproject/gramine/issues/21 above.

> 1. Why does the child fail to receive IPC from the parent suddenly although the parent is still active?

Are you sure the parent is still alive (or rather the IPC worker thread in the parent, but killing it is the last thing process does)?

> 2. Need to debug this error message “Signal 17 queue of the process is full, dropping incoming signal” as this could be the reason the parent failed in the first place.

It's not an error, it's a debug message. You cant have more than 1 instance of a normal signal queue, and in this test multiple children exit - and generate SIGCHILD (17) signal - at once.

**Comment by vijaydhanraj** [comment]
>To be specific: the part about IPC message receiving failure. Are you sure it happens while IPC leader is still alive?

@boryspoplawski I am still trying to understand the IPC code flows and debugging based on the error logs. So, my analysis might not be correct. But from the logs, we see that `T24 (PID: 39264)` thinks the IPC leader is dead but the leader (PID: 39170) still continues to receive messages from that child. Please see lines `3795` and `3828` and from the log snippet below, (sorry I am attaching a picture, to show the line numbers)

![image](https://user-images.githubusercontent.com/37005992/126687245-31e5b44a-27ed-4bb5-9cee-ebc92260289b.png)

And the leader continues with the following error message before exiting,
```
3827 [P39170:shim] debug: IPC worker: received IPC message from 39264: code=1 size=54 seq=0
3828 [P39170:shim] debug: IPC callback from 39264: IPC_MSG_CHILDEXIT(1, 24, 0, 15)
3829 [P39170:shim] error: Unknown process sent a child-death notification: pid: 24, vmid: 39264
3830 [P39170:shim] error: IPC worker: error running IPC callback 1: -22
3831 debug: DkProcessExit: Returning exit code 1
```
>It's not an error, it's a debug message. You cant have more than 1 instance of a normal signal queue, and in this test multiple children exit - and generate SIGCHILD (17) signal - at once.

Got it, thanks!

**Comment by boryspoplawski** [comment]
There seem to be a problem somewhere indeed. The interesting part of the log starts before the part on the screenshot you posted (would be nice to see if IPC leader tried to send a response).
Actually, could you please try the following patch and see if it changes anything:
```
diff --git a/LibOS/shim/src/ipc/shim_ipc.c b/LibOS/shim/src/ipc/shim_ipc.c
index 9e5d6ab3..45572b2e 100644
--- a/LibOS/shim/src/ipc/shim_ipc.c
+++ b/LibOS/shim/src/ipc/shim_ipc.c
@@ -281,6 +281,10 @@ int ipc_send_msg_and_get_response(IDTYPE dest, struct shim_ipc_msg* msg, void**
         goto out;
     }
 
+    lock(&g_msg_waiters_tree_lock);
+    avl_tree_delete(&g_msg_waiters_tree, &waiter.node);
+    unlock(&g_msg_waiters_tree_lock);
+
     if (!waiter.response_data) {
         log_warning("IPC recipient %u died while we were waiting for a message response", dest);
         ret = -ESRCH;
@@ -292,6 +296,8 @@ int ipc_send_msg_and_get_response(IDTYPE dest, struct shim_ipc_msg* msg, void**
         }
         ret = 0;
     }
+    free(waiter.response_data);
+    return ret;
 
 out:
     lock(&g_msg_waiters_tree_lock);
```

**Comment by vijaydhanraj** [comment]
>would be nice to see if IPC leader tried to send a response

Do you mean after one of the children posts this message `[P39264:T24:waitpid03] warning: IPC recipient *39170* died while we were waiting for a message response`? If so no, but I have attached a new log that was taken based on the patch which shows the leader sending an IPC message even after one of the children posts that the leader has died. Please see lines `3404` and `3409`

In this new repro, the leader seems to die very early. We see T13 posting this message. But I have seen similar failure even without the patch.

[repro2_run_waitpid03.txt](https://github.com/oscarlab/graphene/files/6866812/repro2_run_waitpid03.txt)

**Comment by boryspoplawski** [comment]
3377 is the first line with non-normal log entry for `P12542` process. I have no idea what's failing, but this might be the case of #2532. To verify that you could apply this patch:
```
diff --git a/Pal/src/host/Linux-SGX/sgx_exception.c b/Pal/src/host/Linux-SGX/sgx_exception.c
index 62392d3a..2ae75aca 100644
--- a/Pal/src/host/Linux-SGX/sgx_exception.c
+++ b/Pal/src/host/Linux-SGX/sgx_exception.c
@@ -192,6 +192,7 @@ static void handle_async_signal(int signum, siginfo_t* info, struct ucontext* uc
      * was interrupted by calling sgx_entry_return(syscall_return_value=-EINTR, event) */
     /* TODO: we abandon PAL state here (possibly still holding some locks, etc) and return to
      *       enclave; ideally we must unwind/fix the state and only then jump into enclave */
+    log_error("Injecting -EINTR at 0x%lx !!!", ucontext_get_ip(uc));
     ucontext_set_function_parameters(uc, sgx_entry_return, -EINTR, event);
 }
```
and then check if there is this new error log somewhere before the one that fails (`error: IPC worker: receiving message header from 12495 failed: -13`) and that the reported injection address is inside an read ocall, somewhere after a `syscall` instruction.

**Comment by vijaydhanraj** [comment]
It looks like we do see this debug message even in the normal/working case. But I have attached a log `repro3_waitpid.txt` with the issue repro where we do see the debug message being printed before the failure. Do we put this issue on hold until other IPC issues are resolved and then revisit?

[repro3_waitpid.txt](https://github.com/oscarlab/graphene/files/6870721/repro3_waitpid.txt)

**Comment by boryspoplawski** [comment]
> It looks like we do see this debug message even in the normal/working case.

That's expected.

The idea was to check where the address that this log message print is in the failing run (what instruction in which function).

**Comment by vijaydhanraj** [comment]
OK, hope that helps. Just to clarify, I did observe the same address being printed even for a successful case of terminating a thread, like T7, and also observed the same address even for failure case, i.e. when terminating T19.

**Comment by boryspoplawski** [comment]
I would need the untrusted PAL binary to be able to check what address is that (elf base address would be helpful too). The same address can happen, e.g. it's just after syscall instruction in read ocall, and in successful case it's e.g. some random read that's not mandatoriy. Who knows. Needs more debugging/details.

**Comment by vijaydhanraj** [comment]
Debugged `waitpid03` test further with @boryspoplawski and this is indeed the case of #2532 but just in this case the error happens in the `ocall_write` flow. From the logs we can see that the reported injection address is inside `sgx_ocall_write` ocall just after a syscall instruction. 
```
000000000000291a <sgx_ocall_write>:
    291a:   55                      push   %rbp   
    291b:   48 89 e5                mov    %rsp,%rbp
    291e:   48 89 7d c8             mov    %rdi,-0x38(%rbp)
    2922:   48 8b 45 c8             mov    -0x38(%rbp),%rax
    2926:   48 89 45 d0             mov    %rax,-0x30(%rbp)
    292a:   48 8b 45 d0             mov    -0x30(%rbp),%rax
    292e:   8b 40 0c                mov    0xc(%rax),%eax
    2931:   89 c0                   mov    %eax,%eax
    2933:   48 89 45 d8             mov    %rax,-0x28(%rbp)
    2937:   48 8b 45 d0             mov    -0x30(%rbp),%rax
    293b:   48 8b 40 04             mov    0x4(%rax),%rax
    293f:   48 89 45 e0             mov    %rax,-0x20(%rbp)
    2943:   48 8b 45 d0             mov    -0x30(%rbp),%rax
    2947:   8b 00                   mov    (%rax),%eax
    2949:   48 98                   cltq
    294b:   48 89 45 e8             mov    %rax,-0x18(%rbp)
    294f:   48 8b 55 d8             mov    -0x28(%rbp),%rdx
    2953:   48 8b 75 e0             mov    -0x20(%rbp),%rsi
    2957:   48 8b 7d e8             mov    -0x18(%rbp),%rdi
    295b:   b8 01 00 00 00          mov    $0x1,%eax
    2960:   0f 05                   syscall 
    2962:   48 89 45 f0             mov    %rax,-0x10(%rbp) =================>reported injection address
    2966:   48 8b 45 f0             mov    -0x10(%rbp),%rax
    296a:   48 89 45 f8             mov    %rax,-0x8(%rbp)
    296e:   48 8b 45 f8             mov    -0x8(%rbp),%rax
    2972:   5d                      pop    %rbp   
    2973:   c3 
```

As mentioned in #2532, once `-EINTR` is returned, the `TLS` connection gets out of sync and failure ensues.

**Comment by boryspoplawski** [comment]
@vijaydhanraj could you check if #2602 fixes the issue?

**Comment by vijaydhanraj** [comment]
Ran the waitpid03 LTP test over 50 iterations and couldn't repro the issue. @boryspoplawski #2602 does seem to fix the issue.

**Comment by boryspoplawski** [comment]
Good, closing this then as fixed.

**Comment by anjalirai-intel** [comment]
@boryspoplawski 
In the latest jenkins nightly, fcntl12 or fcntl12_64 is still failing with IPC error. The issue still exists

Standard Output
```
fcntl12     0  TINFO  :  Test for errno EMFILE
fcntl12     1  TPASS  :  block 1 PASSED
```

Standard Error
```
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
[P2:T2:fcntl12_64] error: Failed to send IPC msg to 1: -32
```

**Comment by boryspoplawski** [comment]
And is already tracked by https://github.com/gramineproject/gramine/issues/21.

**Comment by boryspoplawski** [comment]
I've re-verified this and the test passes successfully (it prints an error message - tracked by https://github.com/gramineproject/gramine/issues/21 - but it does not affect the test).
